samza-kv-remote/src/main/java/org/apache/samza/table/remote/couchbase/BaseCouchbaseTableFunction.java

```diff
@@ -0,0 +1,212 @@
+  protected String sslKeystoreFile = null;
+  protected String sslKeystorePassword = null;
+  protected String sslTruststoreFile = null;
+  protected String sslTruststorePassword = null;
+  protected int bootstrapCarrierDirectPort = -1;
+  protected int bootstrapCarrierSslPort = -1;
+  protected int bootstrapHttpDirectPort = -1;
+  protected int bootstrapHttpSslPort = -1;
+
+  public BaseCouchbaseTableFunction() {
+  }
+
+  //TODO maybe we can create a builder class to create both read and write functions for the same bucket so that users don't need to type in the same things twice
+  public BaseCouchbaseTableFunction(Class<V> valueClass) {
+    this.valueClass = valueClass;
+  }
+
+  @Override
+  public void init(Context context) {
+    //TODO validation
+    DefaultCouchbaseEnvironment.Builder envBuilder = new DefaultCouchbaseEnvironment.Builder();
+    // ssl settings
+    envBuilder.sslEnabled(sslEnabled).certAuthEnabled(certAuthEnabled);
+    if (sslKeystoreFile != null) {
+      envBuilder.sslKeystoreFile(sslKeystoreFile);
+    }
+    if (sslKeystorePassword != null) {
+      envBuilder.sslKeystorePassword(sslKeystorePassword);
+    }
+    if (sslTruststoreFile != null) {
+      envBuilder.sslTruststoreFile(sslTruststoreFile);
+    }
+    if (sslTruststorePassword != null) {
+      envBuilder.sslTruststorePassword(sslTruststorePassword);
+    }
+    if (bootstrapCarrierDirectPort != -1) {
+      envBuilder.bootstrapCarrierDirectPort(bootstrapCarrierDirectPort);
+    }
+    if (bootstrapCarrierSslPort != -1) {
+      envBuilder.bootstrapCarrierSslPort(bootstrapCarrierSslPort);
+    }
+    if (bootstrapHttpDirectPort != -1) {
+      envBuilder.bootstrapHttpDirectPort(bootstrapHttpDirectPort);
+    }
+    if (bootstrapHttpSslPort != -1) {
+      envBuilder.bootstrapHttpSslPort(bootstrapHttpSslPort);
+    }
+    env = envBuilder.build();
+    cluster = CouchbaseCluster.create(env, clusterNodes);
+    if (username != null && password != null) {
+      cluster.authenticate(username, password);
+    }
+    bucket = cluster.openBucket(bucketName);
+  }
+
+  @Override
+  public void close() {
+    bucket.close();
+    cluster.disconnect();
+    env.shutdown();
+  }
+
+  public <T extends BaseCouchbaseTableFunction<V>> T withTimeout(long timeout, TimeUnit timeUnit) {
+    this.timeout = timeout;
+    this.timeUnit = timeUnit;
+    return (T) this;
+  }
+
+  public <T extends BaseCouchbaseTableFunction<V>> T withTtl(int ttl) {
+    this.ttl = ttl;
+    return (T) this;
+  }
+
+  public <T extends BaseCouchbaseTableFunction<V>> T withSerde(Serde<V> valueSerde) {
+    this.valueSerde = valueSerde;
+    return (T) this;
+  }
+
+  public <T extends BaseCouchbaseTableFunction<V>> T withClusters(Collection<String> clusters) {
+    this.clusterNodes = ImmutableList.copyOf(clusters);
+    //TODO try to get rid of this type casting
+    return (T) this;
+  }
+
+  public <T extends BaseCouchbaseTableFunction<V>> T withUsername(String username) {
```

Comment:
Should we combine username and password, is there a case one only sets username or password?

Reply:
Good point. Some legacy buckets (version < 5.0) may have their own passwords without usernames. But we don't consider that right now.